### PR TITLE
PEP458: Update TUF repository metadata on project index change

### PR DESCRIPTION
**Pull request by lukpueh**
*EDIT 2024/06/21: update description; drop discussion points (see https://github.com/pypi/warehouse/pull/15815#issuecomment-2182420372, https://github.com/pypi/warehouse/pull/15815#issuecomment-2182426602)*

Part 3 in a series of PRs to integrate Repository Service for TUF (RSTUF) with Warehouse for PEP 458 adoption. Previous PR was https://github.com/pypi/warehouse/pull/15484

This PR adds an async task to the `tuf` subpackage, in order to capture project changes in TUF repository metadata via the RSTUF API. Relevant Warehouse views -- file upload and project management (remove, yank) -- are updated to trigger the task. 

This PR also updates the so far unused `render_simple_detail` function for this purpose.

### CODE REVIEW

File: dev/environment [hunk 1]
```diff
@@ -64,6 +64,9 @@ TWOFACTORMANDATE_AVAILABLE=true
 TWOFACTORMANDATE_ENABLED=true
 OIDC_AUDIENCE=pypi
 
+TUF_RSTUF_API_URL="http://rstuf-api"
+TUF_ENABLED=true
```
Comment (simi) [inline dev/environment:68]: Btw. in RubyGems.org we use `RSTUF_API_URL` and we enable it just by presence. Not sure if worth to keep it similar.

https://github.com/rubygems/rubygems.org/blob/ad1269b77e3ac83edac6a8f399125ba9fec0280d/config/initializers/rstuf.rb#L3C9-L7

File: warehouse/tuf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    return response.json()["data"]["bootstrap"]
+
+
+def _rstuf_wait_for_success(server: str, task_id: str) -> None:
+    """Poll RSTUF status API until async task finishes.
+
+    Raises RuntimeError, if task fails.
+    """
+    while True:
```
Comment (simi) [inline warehouse/tuf/__init__.py:40]: Is this in theory of state not being changed infinite loop?

Comment (lukpueh) [reply warehouse/tuf/__init__.py:40]: Yeah, this is just a stopgap solution. We should definitely do better. See discussion item about async task handling in PR description.

**Comment by lukpueh** [comment]
> **Discussion**
> 
> There are a few open question, for which I'd appreciate input from folks, who are more familiar with Warehouse.
> 
>  ...

Any chance to get some feedback from Warehouse maintainers? Also happy to arrange a call to discuss in person. cc @ewdurbin


On a side-note, we released [v1.0.0](https://github.com/secure-systems-lab/securesystemslib/releases/tag/v1.0.0) of the crypto interface used by RSTUF earlier today, and it includes a "VaultSigner", tailored for signing TUF metadata with a HashiCorp Vault key. :)

**Comment by ewdurbin** [comment]
> * Completeness
>   The PR should trigger TUF metadata updates in all relevant views, but someone needs to confirm. Basically, we need to update TUF metadata whenever a project change would result in a different project simple detail file, e.g. on distribution file upload, file removal, or yank. I wonder, if we could register a db hook for those changes, e.g. after_commit, which would also give us better certainty that the db change isn't rolled back.

It may be worth considering the machinery that handles our CDN cache purges (see warehouse/cache/origin) though it is admittedly opaque and a little convoluted to store "purge keys" on the request for handling after flush. This is woefully undocumented, but has been functionally reliable for a long time.

> * Async task handling
>   Metadata update tasks in RSTUF are async. This PR busy polls the RSTUF task status API until the corresponding metadata update task has finished. I'm sure, Warehouse has better ways of dealing with async tasks.

Yes, look for usage of the @tasks.task decorator throughout the codebase.

**Comment by lukpueh** [comment]
Just rebased on latest main, resolving conflicts and including some minor refactorings in the original commits for consistency with https://github.com/pypi/warehouse/pull/16098. More significant changes/additions are added on top. The commit history is now a bit evolutionary. Happy to rewrite/squash, if that's a requirement.

cc @ofek, @trishankatdatadog, @dstufft

**Comment by lukpueh** [comment]
> > * Completeness
> >   The PR should trigger TUF metadata updates in all relevant views, but someone needs to confirm. Basically, we need to update TUF metadata whenever a project change would result in a different project simple detail file, e.g. on distribution file upload, file removal, or yank. I wonder, if we could register a db hook for those changes, e.g. after_commit, which would also give us better certainty that the db change isn't rolled back.
> 
> It may be worth considering the machinery that handles our CDN cache purges (see warehouse/cache/origin) though it is admittedly opaque and a little convoluted to store "purge keys" on the request for handling after flush. This is woefully undocumented, but has been functionally reliable for a long time.

Thanks for the pointer, @ewdurbin! I decided to leave the explicit invocations for now. We can improve this in a follow-up PR.

> 
> > * Async task handling
> >   Metadata update tasks in RSTUF are async. This PR busy polls the RSTUF task status API until the corresponding metadata update task has finished. I'm sure, Warehouse has better ways of dealing with async tasks.
> 
> Yes, look for usage of the @tasks.task decorator throughout the codebase.

Thanks, this worked nicely.

**Comment by lukpueh** [comment]
> * Error handling
>   This means setting up proper retry mechanism for temporary errors, and figuring out user feedback and recovery for permanent errors.

Now that the RSTUF API is called in an async task errors shouldn't affect the normal operation of Warehouse. I suggest to take the PR as is and improve error handling (ignore, retry or notify) in a next iteration.

**Comment by lukpueh** [comment]
> > > * Completeness
> > >   The PR should trigger TUF metadata updates in all relevant views, but someone needs to confirm. Basically, we need to update TUF metadata whenever a project change would result in a different project simple detail file, e.g. on distribution file upload, file removal, or yank. I wonder, if we could register a db hook for those changes, e.g. after_commit, which would also give us better certainty that the db change isn't rolled back.
> > 
> > 
> > It may be worth considering the machinery that handles our CDN cache purges (see warehouse/cache/origin) though it is admittedly opaque and a little convoluted to store "purge keys" on the request for handling after flush. This is woefully undocumented, but has been functionally reliable for a long time.
> 
> Thanks for the pointer, @ewdurbin! I decided to leave the explicit invocations for now. We can improve this in a follow-up PR.
> 
> > > * Async task handling
> > >   Metadata update tasks in RSTUF are async. This PR busy polls the RSTUF task status API until the corresponding metadata update task has finished. I'm sure, Warehouse has better ways of dealing with async tasks.
> > 
> > 
> > Yes, look for usage of the @tasks.task decorator throughout the codebase.
> 
> Thanks, this worked nicely.

**Comment by lukpueh** [comment]
(sorry for the noise, I accidentally hit the "close with comment" button)

**Comment by trishankatdatadog** [comment]
> Just rebased on latest main, resolving conflicts and including some minor refactorings in the original commits for consistency with #16098. More significant changes/additions are added on top. The commit history is now a bit evolutionary. Happy to rewrite/squash, if that's a requirement.
> 
> cc @ofek, @trishankatdatadog, @dstufft

Sorry, was on PTO. How can we best help here? For example, would you like us (esp. Donald) to review?

**Comment by lukpueh** [comment]
Thanks, @trishankatdatadog! A review by anyone would be amazing. Donald seems to be a great candidate because he  knows Warehouse and TUF extremely well.

As said above, I think, this PR can be merged as is. However, I'd appreciate help with improvements in follow-up PRs (see my comments above about [completeness](https://github.com/pypi/warehouse/pull/15815#issuecomment-2182420372) and [error handling](https://github.com/pypi/warehouse/pull/15815#issuecomment-2182426602)).

**Review by firefly9966 — APPROVED** [review]

File: warehouse/manage/views/__init__.py [hunk 3]
```diff
@@ -1861,6 +1862,8 @@ def delete_project(project, request):
 
     remove_project(project, request)
 
+    request.task(tuf.update_metadata).delay(project.id)
```
Comment (trishankatdatadog) [inline warehouse/manage/views/__init__.py:1865]: I see these calls a fair bit. Is there anything we could automatically do to reduce the burden of the user?

File: tests/unit/packaging/test_utils.py [hunk 4]
```diff
@@ -90,7 +91,7 @@ def test_render_simple_detail_with_store(db_request, monkeypatch, jinja):
 
     fake_named_temporary_file = pretend.stub(
         name="/tmp/wutang",
-        write=pretend.call_recorder(lambda data: None),
+        write=pretend.call_recorder(lambda data: 42),
```
Comment (trishankatdatadog) [inline tests/unit/packaging/test_utils.py:94]: Nitpick: could we specify and reuse a constant for the size (42)?

**Review by trishankatdatadog — COMMENTED** [review]
LGTM FWIW!

My high-level Q is: does this work with the JSON Simple API (particularly for [PEP 740](https://peps.python.org/pep-0740/#json-based-simple-api))?

File: warehouse/manage/views/__init__.py (hunk 3 above)
Comment (lukpueh) [reply warehouse/manage/views/__init__.py:1865]: See discussion about completeness above: https://github.com/pypi/warehouse/pull/15815#issuecomment-2182420372

TLDR: This should be improved, but could happen in a follow-up PR.

**Comment by lukpueh** [comment]
Thanks for the review, @trishankatdatadog!

> My high-level Q is: does this work with the JSON Simple API (particularly for [PEP 740](https://peps.python.org/pep-0740/#json-based-simple-api))?

Great catch! This PR only covers the html simple index, but could be easily extended to also include the json simple index. 

AFAICS we'd *only* have to make sure that
- project json simple index is persistent (for some time) like the [html simple index is](https://github.com/pypi/warehouse/blob/cda5311284b7d23ca59ce26d970b464c5515e5a5/warehouse/packaging/utils.py#L95), and
- post [two TUF targets instead of one](https://github.com/pypi/warehouse/blob/cda5311284b7d23ca59ce26d970b464c5515e5a5/warehouse/tuf/__init__.py#L126-L132).

[PEP 691](https://peps.python.org/pep-0691/#tuf-support-pep-458) that TUF uses these target paths:
- `simple/PROJECT/vnd.pypi.simple.v1.html`
- `simple/PROJECT/vnd.pypi.simple.v1.json`

**Comment by trishankatdatadog** [comment]
> AFAICS we'd _only_ have to make sure that

Got it, thanks. @woodruffw is this correct: did we miss anything?

File: warehouse/tuf/__init__.py [hunk 5]
```diff
@@ -72,3 +101,38 @@ def wait_for_success(server: str, task_id: str):
...
+    server = request.registry.settings["tuf.rstuf_api_url"]
+    if not server:
+        return
+
+    project = request.db.query(Project).filter(Project.id == project_id).one()
+
+    # NOTE: We ignore the returned simple detail path with the content hash as
+    # infix. In TUF metadata the project name and hash are listed separately, so
+    # that there is only one entry per target file, even if the content changes.
+    digest, _, size = render_simple_detail(project, request, store=True)
```
Comment (miketheman) [inline warehouse/tuf/__init__.py:123]: If we explicitly ignore the simple_detail_path, why should the function continue to return it at all? Does any other caller use it?

File: warehouse/packaging/utils.py [hunk 6]
```diff
@@ -107,12 +107,12 @@ def render_simple_detail(project, request, store=False):
...
     )
 
-    if store:
-        storage = request.find_service(ISimpleStorage)
-        with tempfile.NamedTemporaryFile() as f:
-            f.write(content.encode("utf-8"))
-            f.flush()
+    with tempfile.NamedTemporaryFile() as f:
+        simple_detail_size = f.write(content.encode("utf-8"))
+        f.flush()
```
Comment (miketheman) [inline warehouse/packaging/utils.py:112]: Is this the only way for us to measure the size by writing out to disk? Could we keep the writing to disk still behind the `if store` conditional, and calculate the size in memory?

File: warehouse/manage/views/__init__.py (hunk 3 above)
Comment (miketheman) [reply warehouse/manage/views/__init__.py:1865]: I'll echo this - we don't have many tasks directly called inside views today, rather we listen to the objects being changed via sqlalchemy and trigger tasks that way.

Here's a couple of examples:

https://github.com/pypi/warehouse/blob/e8195e40986d63f9b72926a67eb6acfd43b36932/warehouse/search/__init__.py#L27-L64

https://github.com/pypi/warehouse/blob/e8195e40986d63f9b72926a67eb6acfd43b36932/warehouse/observations/tasks.py#L33-L54

**Review by miketheman — DISMISSED** [review]
Hey gang! I reviewed, and commented on how this is constructed, let me know if anything isn't clear!

**Review by miketheman — COMMENTED** [review]
Whoops, clicked the wrong button

**Review by miketheman — COMMENTED** [review]
I can't seem to use GitHub correctly, gonna stop for today 😆

**Comment by lukpueh** [comment]
Thanks for the feedback, @miketheman!

@kairoaraujo, @trishankatdatadog, and friends, please feel free to take over as discussed the other day!

**Comment by kairoaraujo** [comment]
Note: I pushed some comment fixes, but I am still not finished (Work in Progress)

File: dev/environment [hunk 7]
```diff
@@ -64,6 +64,8 @@ TWOFACTORMANDATE_AVAILABLE=true
 TWOFACTORMANDATE_ENABLED=true
 OIDC_AUDIENCE=pypi
 
+TUF_RSTUF_API_URL="http://rstuf-api"
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply dev/environment]: Solved with commit  `refactor: remove prefixes tuf/TUF from config`

File: warehouse/tuf/__init__.py [hunk 8]
```diff
@@ -72,3 +101,38 @@ def wait_for_success(server: str, task_id: str):
 
     else:
         raise RSTUFError("RSTUF job failed, please check payload and retry")
+
+
+@tasks.task(ignore_result=True, acks_late=True)
+def update_metadata(request: Request, project_id: UUID):
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply warehouse/tuf/__init__.py]: Solved with commit `refactor: split tuf.__init__.tasks to tuf.tasks`

File: warehouse/config.py [hunk 9]
```diff
@@ -367,6 +367,7 @@ def configure(settings=None):
         "OIDC_BACKEND",
         default="warehouse.oidc.services.OIDCPublisherService",
     )
+    maybe_set(settings, "tuf.rstuf_api_url", "TUF_RSTUF_API_URL")
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply warehouse/config.py]: Solved with commit  `refactor: remove prefixes tuf/TUF from config`

File: warehouse/forklift/legacy.py [hunk 10]
```diff
@@ -1168,6 +1169,8 @@ def file_upload(request):
 
     request.db.flush()  # flush db now so server default values are populated for celery
 
+    request.task(tuf.update_metadata).delay(release.project.id)
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply warehouse/forklift/legacy.py]: Solved with commit `chore: wrapped metadata update if rstuf is enabled`

File: tests/unit/tuf/test_tuf.py [hunk 11]
```diff
@@ -88,3 +119,55 @@ def test_wait_for_success_error(self, state, iterations, monkeypatch):
...
+        project = stub(normalized_name=project_name)
+
+        one = call_recorder(lambda: project)
+        db_request.db.query = lambda a: stub(filter=lambda a: stub(one=one))
+
+        # Test early return, if no RSTUF API URL configured
+        db_request.registry.settings = {"tuf.rstuf_api_url": None}
+        tuf.update_metadata(db_request, project_id)
+
+        assert not one.calls
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply tests/unit/tuf/test_tuf.py]: Solved with commit `refactor: split tuf.__init__.tasks to tuf.tasks`

File: warehouse/tuf/__init__.py
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply warehouse/tuf/__init__.py:1]: Done with commit `chore: Implement TUF Interface and RSTUF Service`

File: tests/unit/tuf/test_tuf.py [hunk 12]
```diff
@@ -57,6 +57,37 @@ def test_post_bootstrap(self, monkeypatch):
...
+        # Test success
+        result = tuf.post_artifacts(self.server, payload)
+
+        assert result == self.task_id
+        assert post.calls == [call(f"{self.server}/api/v1/artifacts", json=payload)]
+
+        # Test fail with incomplete response json (i.e. no bootstrap error)
+        del resp_json["data"]
+        with pytest.raises(tuf.RSTUFNoBootstrapError):
+            tuf.post_artifacts(self.server, payload)
```
[the remark this answers is not in the document]
Comment (kairoaraujo) [reply tests/unit/tuf/test_tuf.py]: Hi @miketheman,

It has changed with the refactoring I did, but should I create two separated tests?

https://github.com/pypi/warehouse/blob/86803ace65aea255f5639cf3321c175a49ee6940/tests/unit/tuf/test_services.py#L127

File: warehouse/packaging/utils.py (hunk 6 above)
Comment (kairoaraujo) [reply warehouse/packaging/utils.py]: Addressed in `refactor: utils.render_simple_detail`

File: warehouse/tuf/__init__.py (hunk 5 above)
Comment (kairoaraujo) [reply warehouse/tuf/__init__.py]: Removed in the commit `refactor: utils.render_simple_detail`

**Comment by kairoaraujo** [comment]
Hey folks,
Ready for another review! 🎉

File: warehouse/tuf/tasks.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    are enough, as they already include all relevant distribution file infos.
+    """
+    rstuf_service = request.find_service(ITUFService, context=None)
+
+    project = request.db.query(Project).filter(Project.id == project_id).one()
+
+    # NOTE: We ignore the returned simple detail path with the content hash as
+    # infix. In TUF metadata the project name and hash are listed separately, so
+    # that there is only one entry per target file, even if the content changes.
+    digest, size = render_simple_detail(project, request, store=True)
```
Comment (woodruffw) [inline warehouse/tuf/tasks.py:37]: How does this interact with/handle the other artifacts/pieces of metadata produced on file upload? In particular, I'm thinking of:

1. The JSON (PEP 691) simple index
2. The detached metadata (PEP 658)

File: warehouse/tuf/services.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                case "ERRORED" | "REVOKED" | "REJECTED":
+                    raise RSTUFError(
+                        "RSTUF internal problem, please check RSTUF health"
+                    )
+
+                case _:
+                    raise RSTUFError(f"RSTUF job returned unexpected state: {state}")
+
+        else:
+            raise RSTUFError("RSTUF job failed, please check payload and retry")
```
Comment (woodruffw) [inline warehouse/tuf/services.py:85]: For my own elucidation: is there a consistency risk with TUF if a particular upload fails and the next sequenced one succeeds? IIRC there isn't (since TUF isn't trying to impose a canonical ordering of uploads), but I wanted to make sure I'm not missing something there 🙂

Comment (woodruffw) [reply warehouse/tuf/services.py:85]: (Separate from the more basic consistency risk of not retrying, that is.)
